pkg/ingester/ingester_v2.go

```diff
@@ -240,6 +218,78 @@ func (i *Ingester) v2LabelNames(ctx old_ctx.Context, req *client.LabelNamesReque
 	}, nil
 }
 
+func (i *Ingester) v2MetricsForLabelMatchers(ctx old_ctx.Context, req *client.MetricsForLabelMatchersRequest) (*client.MetricsForLabelMatchersResponse, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	db := i.getTSDB(userID)
+	if db == nil {
+		return &client.MetricsForLabelMatchersResponse{}, nil
+	}
+
+	// Parse the request
+	from, to, matchersSet, err := client.FromMetricsForLabelMatchersRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new instance of the TSDB querier
+	q, err := db.Querier(int64(from), int64(to))
+	if err != nil {
+		return nil, err
+	}
+	defer q.Close()
+
+	// Run a query for each matchers set and collect all the results
+	added := model.FingerprintSet{}
```

Comment:
I could fix it here, but wouldn't be enough, because the result of this method is then merged by the [distributor](https://github.com/cortexproject/cortex/blob/master/pkg/distributor/distributor.go#L524) which uses a fingerprint-based map to build the result. I believe this fingerprint hash collision issue should be discussed and fixed in a separate PR, cause it's quite wide spread in Cortex.

Reply:
> I believe this fingerprint hash collision issue should be discussed and fixed in a separate PR, cause it's quite wide spread in Cortex.

It's actually pretty well handled elsewhere in the Cortex, except one small bug when fetching labels. I'll take a look at what you're pointing to to understand your comment, but on the first sight, this code introduces new bug.

Reply:
I've switched this specific part to a map of strings, avoiding to use any hashing at all. What's your take?

Reply:
👍  Thanks!